examples/userapi/04_boundary_conditions.ipynb

```diff
@@ -0,0 +1,530 @@
+    "          subdomain=grid.subdomains['main'])\n",
+    "\n",
+    "eq_p = Eq(p.forward, p - dt*velocity**2*density*div(v.forward),\n",
+    "           subdomain=grid.subdomains['main'])"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We will also need to set up `p_i` to calculate the integral of `p` over time for out PMLs:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 12,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "eq_p_i = Eq(p_i.forward, p_i + dt*(p.forward+p)/2)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "And add the equations for our damped region:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 13,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Left side\n",
+    "eq_v_damp_left = Eq(v.forward,\n",
+    "                    (1-d_l)*v - dt*grad(p)/density,\n",
+    "                    subdomain=grid.subdomains['left'])\n",
+    "\n",
+    "eq_p_damp_left = Eq(p.forward,\n",
+    "                    (1-gamma*velocity**2*dt-d_l*dt)*p\n",
+    "                    - d_l*gamma*velocity**2*p_i\n",
+    "                    - dt*velocity**2*density*div(v.forward),\n",
+    "                    subdomain=grid.subdomains['left'])\n",
+    "\n",
+    "# Right side\n",
+    "eq_v_damp_right = Eq(v.forward,\n",
+    "                     (1-d_r)*v - dt*grad(p)/density,\n",
+    "                     subdomain=grid.subdomains['right'])\n",
+    "\n",
+    "eq_p_damp_right = Eq(p.forward,\n",
+    "                     (1-gamma*velocity**2*dt-d_r*dt)*p\n",
+    "                     - d_r*gamma*velocity**2*p_i\n",
+    "                     - dt*velocity**2*density*div(v.forward),\n",
+    "                     subdomain=grid.subdomains['right'])\n",
+    "\n",
+    "# Base edge\n",
+    "eq_v_damp_base = Eq(v.forward,\n",
+    "                    (1-d_b)*v - dt*grad(p)/density,\n",
+    "                    subdomain=grid.subdomains['base'])\n",
+    "\n",
+    "eq_p_damp_base = Eq(p.forward,\n",
+    "                    (1-gamma*velocity**2*dt-d_b*dt)*p\n",
+    "                    - d_b*gamma*velocity**2*p_i\n",
+    "                    - dt*velocity**2*density*div(v.forward),\n",
+    "                    subdomain=grid.subdomains['base'])"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Add our free surface boundary conditions:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 14,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "time = grid.stepping_dim\n",
+    "\n",
+    "bc_p = [Eq(p[time+1, x, 3], 0.)]  # Set pressure field to zero at boundary\n",
+    "\n",
+    "bc_v = [Eq(v[1][time+1, x, 2], v[1][time+1, x, 4])]  # Mirror y velocity across boundary\n",
+    "bc_v += [Eq(v[1][time+1, x, 1], v[1][time+1, x, 5])]\n",
+    "bc_v += [Eq(v[1][time+1, x, 0], v[1][time+1, x, 6])]"
```

Comment:
THis is basically hard-coded for space_order=6. Would be good to use a SubDim/SubDomain to make it generic and be able to switch space order easily.

see for example https://github.com/devitocodes/devito/blob/8fb68519d477e12df14960264980c3ffb3b526f9/examples/seismic/boundaries.py#L57

Reply:
Also, this is supposed to be opposite sign 
 [Eq(v[1][time+1, x, 1], -v[1][time+1, x, 5])]